C error on generic anon fn call

### Describe the bug

Running `./v -g -o vdbg cmd/v && ./vdbg 'generic_anon_fn.v' && 'generic_anon_fn'` on `generic_anon_fn.v` produced a compiler error.

### Reproduction Steps

```sh
./v -g -o vdbg cmd/v && ./vdbg 'generic_anon_fn.v' && 'generic_anon_fn'
```

```v
fn foo[T](do fn () !T) !T {
	mut r := T{}
	r = do()!
	return r
}

fn main() {
	do := fn () !int {
		return -1
	}
	r := foo(do)!
	assert r == -1
}
```

Note that to anon fn `fn () !int` is assigned to variable named `do`, `foo()` calls the local variable of the same name as fn `r = do()!`. If you change `do` variable name another cgen error is ocurred — https://github.com/vlang/v/issues/27108

### Expected Behavior

Successful compilation.

### Current Behavior

```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/generic_anon_fn.01KR1TVGNWM7YEYBPKH18YD5X6.tmp.c:8527: error: '_v_do' undeclared
=================================================================
Try passing `-g` when compiling, to see a .v file:line information, that correlates more with the C error.
(Alternatively, pass `-show-c-output`, to print the full C error message).
builder error: 
==================
C error found while compiling generated C code.
This can be caused by invalid C interop code, C compiler flags, or a V compiler bug.
If your code is pure V and this still happens, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

### Possible Solution

_No response_

### Additional Information/Context

Generated by `v bug` from `generic_anon_fn.v`.

### V version

V 0.5.1 701d7f3

### Environment details (OS name and version, etc.)

```
|V full version      |V 0.5.1 2f3c0f1f397c08bfc5ef73e7ff62794471b2b811.701d7f3
|:-------------------|:-------------------
|OS                  |linux, Linux version 7.0.3-arch1-1 (linux@archlinux) (gcc (GCC) 15.2.1 20260209, GNU ld (GNU Binutils) 2.46) #1 SMP PREEMPT_DYNAMIC Thu, 30 Apr 2026 18:41:12 +0000
|Processor           |16 cpus, 64bit, little endian, 12th Gen Intel(R) Core(TM) i5-1240P
|Memory              |4.04GB/15.34GB
|                    |
|V executable        |/home/ge/.vlang/v
|V last modified time|2026-05-07 18:19:14
|                    |
|V home dir          |OK, value: /home/ge/.vlang
|VMODULES            |OK, value: /home/ge/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /tmp
|                    |
|Git version         |git version 2.54.0
|V git status        |weekly.2026.03-1869-g701d7f37
|.git/config present |true
|                    |
|cc version          |cc (GCC) 15.2.1 20260209
|gcc version         |gcc (GCC) 15.2.1 20260209
|clang version       |clang version 22.1.3
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |N/A
|glibc version       |ldd (GNU libc) 2.43
```

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1811,6 +1811,14 @@ fn (mut g Gen) gen_anon_fn(mut node ast.AnonFn) {
 		mut has_inherited := false
 		mut is_ptr := false
 		var_name := c_name(var.name)
+		// `ref_name` is how the captured variable is referenced in the enclosing
+		// scope. For fn-typed local variables `fn_var_signature` emits `c_fn_name`,
+		// so the RHS reference must match. The struct field name (LHS) stays as
+		// `c_name(var.name)` to match `closure_field_cname` reads inside the body.
+		mut ref_name := var_name
+		if parent_obj := node.decl.scope.parent.find_var(var.name) {
+			ref_name = g.var_cname(parent_obj)
+		}
 		if g.is_mut_closure_fixed_array(var) {
 			if obj := node.decl.scope.find_var(var.name) {
 				if obj.has_inherited {
@@ -1819,7 +1827,7 @@ fn (mut g Gen) gen_anon_fn(mut node ast.AnonFn) {
 				}
 			}
 			if !has_inherited {
-				g.writeln('.${var_name} = ${var_name},')
+				g.writeln('.${var_name} = ${ref_name},')
 			}
 			continue
 		}
@@ -1852,13 +1860,13 @@ fn (mut g Gen) gen_anon_fn(mut node ast.AnonFn) {
 			if var_sym.info is ast.ArrayFixed {
 				g.write('.${var_name} = {')
 				for i in 0 .. var_sym.info.size {
-					g.write('${var_name}[${i}],')
+					g.write('${ref_name}[${i}],')
 				}
 				g.writeln('},')
 			} else if g.is_autofree && !var.is_mut && var_sym.info is ast.Array {
-				g.writeln('.${var_name} = builtin__array_clone(&${var_name}),')
+				g.writeln('.${var_name} = builtin__array_clone(&${ref_name}),')
 			} else if g.is_autofree && !var.is_mut && var_sym.kind == .string {
-				g.writeln('.${var_name} = builtin__string_clone(${var_name}),')
+				g.writeln('.${var_name} = builtin__string_clone(${ref_name}),')
 			} else {
 				mut is_auto_heap := false
 				mut is_auto_deref_capture := false
@@ -1877,9 +1885,9 @@ fn (mut g Gen) gen_anon_fn(mut node ast.AnonFn) {
 					}
 				}
 				if (is_auto_heap && !is_ptr) || is_auto_deref_capture || field_name != '' {
-					g.writeln('.${var_name} = *${var_name}${field_name},')
+					g.writeln('.${var_name} = *${ref_name}${field_name},')
 				} else {
-					g.writeln('.${var_name} = ${var_name},')
+					g.writeln('.${var_name} = ${ref_name},')
 				}
 			}
 		}

--- a/vlib/v/gen/c/utils.v
+++ b/vlib/v/gen/c/utils.v
@@ -1760,11 +1760,11 @@ fn (mut g Gen) unwrap(typ ast.Type) Type {
 // generate function variable definition, e.g. `void (*var_name) (int, string)`
 fn (mut g Gen) fn_var_signature(var_type ast.Type, return_type ast.Type, arg_types []ast.Type, var_name string) string {
 	if var_type.has_flag(.option) || var_type.has_flag(.result) {
-		return '${g.styp(var_type)} ${c_name(var_name)}'
+		return '${g.styp(var_type)} ${c_fn_name(var_name)}'
 	}
 	ret_styp := g.styp(return_type)
 	nr_muls := var_type.nr_muls()
-	mut sig := '${ret_styp} (${'*'.repeat(nr_muls + 1)}${c_name(var_name)}) ('
+	mut sig := '${ret_styp} (${'*'.repeat(nr_muls + 1)}${c_fn_name(var_name)}) ('
 	for j, arg_typ in arg_types {
 		arg_sym := g.table.sym(arg_typ)
 		if arg_sym.info is ast.FnType {

--- a/vlib/v/tests/fns/fn_var_name_using_reserved_test.v
+++ b/vlib/v/tests/fns/fn_var_name_using_reserved_test.v
@@ -9,3 +9,27 @@ fn test_fn_var_name_using_reserved() {
 	println(new(42))
 	assert new(42) == 42
 }
+
+fn call_fn(do fn () int) int {
+	return do()
+}
+
+// https://github.com/vlang/v/issues/27109
+fn test_reserved_name_as_fn_param_and_call_arg() {
+	do := fn () int {
+		return -1
+	}
+	assert call_fn(do) == -1
+}
+
+fn call_result_fn(do fn () !int) !int {
+	return do()!
+}
+
+fn test_reserved_name_as_result_fn_param_and_call_arg() {
+	do := fn () !int {
+		return -1
+	}
+	r := call_result_fn(do)!
+	assert r == -1
+}

